### DISCUSSION TITLE
Can I use it with Next.js?

### DISCUSSION THREAD

**Discussion by iemong** [edmundhung/conform#355] [Q&A] [answered]
Hi!

Thanks for building this library.
Is it possible to use conform for form validation with server actions in Next.js?

**Comment by edmundhung** [comment, marked as answer]
I have heard people using it with nextjs but I am also not sure if the current version works well with server action yet as it wasn't the main target

Having said that, the upcoming v1 does support server action and should be out in early January.

**Reply by iemong** [reply]
Thank you very much! I'm looking forward to it.

**Reply by deadcoder0904** [reply]
@edmundhung is v1 release near? i do see `^0.9.1` in my `package.json`. i've tried conform in remix & would love a port in next.js.

im currently having an issue with putting something in `lastSubmission` which i've commented & `signup` is getting some typescript error:

```ts
import React from 'react'
import ReactDOM from 'react-dom'
import Link from 'next/link'
import { Argon2id } from 'oslo/password'
import { cookies } from 'next/headers'
import { redirect } from 'next/navigation'
import { generateId } from 'lucia'
import { z } from 'zod'
import { conform, useForm } from '@conform-to/react'
import { getFieldsetConstraint, parse } from '@conform-to/zod'

import { lucia } from '@/app/auth/lucia'
import { db } from '@/app/db/index'
import { userTable } from '@/app/db/schema'

export const schema = z.object({
  email: z
    .string({ required_error: 'Email is required' })
    .email('Invalid email address'),
})

async function signup(formData: FormData) {
  'use server'

  const userId = generateId(15)

  const submission = await parse(formData, {
    schema: schema.transform(async (data, ctx) => {
      const id = generateId(15)
      const session = await db.insert(userTable).values({ id, ...data })

      if (!session) {
        ctx.addIssue({
          path: ['email'],
          code: z.ZodIssueCode.custom,
          message: 'Invalid email or password',
        })
        return z.NEVER
      }
      return { ...data, session }
    }),
    async: true,
  })

  const session = await lucia.createSession(userId, {})
  const sessionCookie = lucia.createSessionCookie(session.id)
  cookies().set(
    sessionCookie.name,
    sessionCookie.value,
    sessionCookie.attributes
  )
  return redirect('/')
}

export const Signup = () => {
  const [formState, action] = ReactDOM.useFormState(signup, null)

  const [form, fields] = useForm({
    id: 'login-form',
    constraint: getFieldsetConstraint(schema),
    // lastSubmission: formState.message,
    onValidate({ formData }) {
      return parse(formData, { schema })
    },
    shouldValidate: 'onBlur',
  })

  return (
    <>
      <Link href="/" className="anchor-dark">
        go home
      </Link>
      <form action={action}>
        <input type="email" name="email" id="email" />
        <button type="submit">signup</button>
      </form>
    </>
  )
}

export default Signup
```

lmk if it wont work so i'll drop it & go for [react-hook-form](https://nehalist.io/react-hook-form-with-nextjs-server-actions/) for now.

**Reply by edmundhung** [reply]
We have released [v1.0.0-rc.0](https://github.com/edmundhung/conform/releases/tag/v1.0.0-rc.0) early this week and working hard on getting all the docs updated. If you don't mind help testing it, feel free to try it out. 

You can find a nextjs example using v1.0.0-rc.0 here: https://codesandbox.io/p/devbox/fast-fast-8jwsxk?file=%2Fapp%2Fform.tsx%3A21%2C31-21%2C43

**Reply by deadcoder0904** [reply]
is there a github repo? the codesandbox is stuck on loading.

i'm close already to get it working. `useFormState` is throwing some weird error when i add an action:

> No overload matches this call.
  Overload 1 of 2, '(action: (state: SubmissionResult<string[]>) => SubmissionResult<string[]> | Promise<SubmissionResult<string[]>>, initialState: SubmissionResult<...>, permalink?: string | undefined): [state: ...]', gave the following error.
    Argument of type '(formData: FormData) => Promise<SubmissionResult<string[]>>' is not assignable to parameter of type '(state: SubmissionResult<string[]>) => SubmissionResult<string[]> | Promise<SubmissionResult<string[]>>'.
      Types of parameters 'formData' and 'state' are incompatible.
        Type 'SubmissionResult<string[]>' is missing the following properties from type 'FormData': append, delete, get, getAll, and 7 more.
  Overload 2 of 2, '(action: (state: SubmissionResult<string[]>, payload: unknown) => SubmissionResult<string[]> | Promise<SubmissionResult<string[]>>, initialState: SubmissionResult<...>, permalink?: string | undefined): [state: ...]', gave the following error.
    Argument of type '(formData: FormData) => Promise<SubmissionResult<string[]>>' is not assignable to parameter of type '(state: SubmissionResult<string[]>, payload: unknown) => SubmissionResult<string[]> | Promise<SubmissionResult<string[]>>'.
      Types of parameters 'formData' and 'state' are incompatible.
        Type 'SubmissionResult<string[]>' is not assignable to type 'FormData'.ts(2769)

i'm sure i can figure it out if i see some example.

**Reply by deadcoder0904** [reply]
okay, codesandbox loaded in another browser. i'll test it.

**Reply by edmundhung** [reply]
Repo: https://github.com/edmundhung/conform/tree/v1/examples/nextjs

**Reply by deadcoder0904** [reply]
thanks, @edmundhung it works fine. just tested it. love it. makes forms easy & fast.
